src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_ucp.h"
+#include "tl_ucp_tag.h"
+
+static void ucc_tl_ucp_req_init(void *request)
+{
+    ucc_tl_ucp_req_t *req = (ucc_tl_ucp_req_t *)request;
+    req->status           = UCC_OPERATION_INITIALIZED;
+}
+
+static void ucc_tl_ucp_req_cleanup(void *request)
+{
+}
+
+UCC_CLASS_INIT_FUNC(ucc_tl_ucp_context_t,
+                    const ucc_base_context_params_t *params,
+                    const ucc_base_config_t *config)
+{
+    ucc_tl_context_config_t *tl_config =
+        ucc_derived_of(config, ucc_tl_context_config_t);
+    ucc_status_t        ucc_status = UCC_OK;
+    ucp_worker_params_t worker_params;
+    ucp_worker_attr_t   worker_attr;
+    ucp_params_t        ucp_params;
+    ucp_config_t       *ucp_config;
+    ucp_context_h       ucp_context;
+    ucp_worker_h        ucp_worker;
+    ucs_status_t        status;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_config->tl_lib);
+    status = ucp_config_read(params->prefix, NULL, &ucp_config);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to read ucp configuration, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg;
+    }
+
+    ucp_params.field_mask =
+        UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE |
+        UCP_PARAM_FIELD_REQUEST_INIT | UCP_PARAM_FIELD_REQUEST_CLEANUP |
+        UCP_PARAM_FIELD_TAG_SENDER_MASK;
+    ucp_params.features        = UCP_FEATURE_TAG | UCP_FEATURE_RMA;
+    ucp_params.request_size    = sizeof(ucc_tl_ucp_req_t);
+    ucp_params.request_init    = ucc_tl_ucp_req_init;
+    ucp_params.request_cleanup = ucc_tl_ucp_req_cleanup;
+    ucp_params.tag_sender_mask = UCC_TL_UCP_TAG_SENDER_MASK;
+
+    if (params->estimated_num_ppn > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_PPN;
+        ucp_params.estimated_num_ppn = params->estimated_num_ppn;
+    }
+
+    if (params->estimated_num_eps > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_EPS;
+        ucp_params.estimated_num_eps = params->estimated_num_eps;
+    }
+
+    status = ucp_init(&ucp_params, ucp_config, &ucp_context);
+    ucp_config_release(ucp_config);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to init ucp context, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg;
+    }
+
+    worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    switch (params->thread_mode) {
+    case UCC_THREAD_SINGLE:
+    case UCC_THREAD_FUNNELED:
+        worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+        break;
+    case UCC_THREAD_MULTIPLE:
+        worker_params.thread_mode = UCS_THREAD_MODE_MULTI;
+        break;
+    default:
+        /* unreachable */
+        ucc_assert(0);
+        break;
+    }
```

Comment:
If UCC context is created in SHARED mode, do you need UCS_THREAD_MODE_MULTIPLE, even in UCC_THREAD_SINGLE ? I don’t for sure either way, but it seems like a corner case that can cause problems. Thoughts?

Reply:
@manjugv again i see that you use "UCC_CONTEXT_SHARED" in a different manner than you described to WG and in the manual. Previously, you were saying the SHARED has nothing to do with threading. It just means that multiple teams can be created from a ucc_context. If user wants to run collectives on those teams in a threaded env user HAS TO set THREAD MULTIPLE.

Plz, clarify again what UCC_CONTEXT_SHARED means?